### DISCUSSION TITLE
Can someone explain the handling of `Tstr_open` in translmod.ml?

### DISCUSSION THREAD

**Discussion by gasche** [ocaml/ocaml#12785] [Q&A]
@lthls and myself are tearing our hair (that's a lot of hair) over the translmod.ml logic for Tstr_open and its relation to the Tstr_include case above.

https://github.com/ocaml/ocaml/blob/1537083e31490e0dbfebaf557ca0f8379099e271/lambda/translmod.ml#L701-L750

The two code look *extremely* similar (where are the differences?), and it looks very much like `Tstr_open` is in fact adding items to the current module, when intuitively -- and in our tests -- that would not happen.

Notice in particular that the "bound value identifiers" of the opened modules are added to the `newfields` accumulator, which intuitively should be the fields in the resulting structure. What are we missing? (Could there be a comment that explains what is going on?)

**Comment by gasche** [comment]
(cc @garrigue and maybe @lpw25, @trefis?)

**Comment by lthls** [comment]
My current guess is that both open and include work the same way. During type-checking, we first add the inner items to the outer structure (unconditionally), and then we use the signature of the final construction to build a coercion that fits the expected signature (for the outer module). The difference is that In the case of open, the items are not added to the signature, while with include they are (at least until they get removed by either shadowing or a restricted signature constraint).

**Reply by lthls** [reply]
If this is correct, I'd appreciate an explicit confirmation so that we can document it properly somewhere.

**Comment by xavierleroy** [comment]
This code looks unfamiliar to me too.  It was introduced  in https://github.com/ocaml/ocaml/commit/97329f30edcf94f98621a200c33e332cef2761e4 "Extend open to arbritrary module expressions in structures and to applicative module paths in signatures" by @objmagic and @trefis.  See also #2147. Before that, the code was as I wrote it in 1996, namely no lambda code was generated for `Tstr_open`.

**Reply by lthls** [reply]
Thanks for finding the PR. I looked at it, in particular at the review comments, and it seems that @lpw25 was as confused as we were by this part of the code but didn't push for a clearer documentation.
What I take from it is that the generalized `open` is handled almost like `include`; I think in hindsight it might have been better to compile `Pstr_open` to `Tstr_include` (with a dedicated flag) when the module expression is not a simple path, and reserve `Tstr_open` for the simple version.

**Reply by gasche** [reply]
For me the question remains of what is the difference in compilation strategy between the two constructs, and where it comes from.

**Comment by yallop** [comment]
The code does look similar when `open_bound_items` is non-empty.  In that case, `open` needs to generate code (consider `open struct let () = exit 0 end`).  When `open_bound_items` is empty, there's no code generated; `transl_structure` just calls itself recursively on the `rem`ainder of the structure items

**Reply by gasche** [reply]
That makes sense, but my question is what makes it so that those items are not added in the output module. If we write

```ocaml
module M_include = struct
  include struct let x = 1 end
end

module M_open = struct
  open struct let x = 1 end
end
```

then (of course) `M_include` has a field/item `x`, and `M_open` has no field/item. But where does this difference come from? In particular, I would expect the "bound idents" of the item to generate code but to *not* be added to the `{,new}fields` accumulator, but they appear to be.

**Comment by gasche** [comment]
Note: typemod.ml carries some `visibility` information that is either `Hidden` or `Exported`, so it is possible that *somewhere* information is added to somehow mark the `open`-added fields/items as "hidden".

As noted by @lthls, another hypothesis is that we pretend that the fields are added, but in fact they are ignored in the end because the module expression is wrapped in a coercion that hides them.

**Comment by lthls** [comment]
I believe I have a satisfying answer to the question. It involves a lot of different pieces of code, I'll try to summarize everything here.
- First, when typing a structure, we always keep all the fields even if they are not exported in the signature. For modules with a restricted signature, this is handled in the obvious way by a `Tmod_constraint` wrapper which introduces a coercion between the concrete signature (of the structure with all its fields) and the declared signature. For cases like `module M = struct let x = 0 let x = "1" end`, it is handled by in implicit constraint (`Tmod_constraint (_, _, Tmodtype_implicit, _)`). The code is there: https://github.com/ocaml/ocaml/blob/cbc5a0c807c77d8fb6d687e36f74af33e4c27bb5/typing/typemod.ml#L2202-L2203
These implicit constraints are not printed with `-dtypedtree`, which made it particularly hard to understand what was going on. But in essence, the structure itself still has a concrete signature containing all fields, even the duplicated ones, and needs a coercion to get back to the actual inferred type.
- For `include`, the implementation is then obvious: just add all fields to the structure, the constraint at the end (implicit or explicit) will take care of filtering the ones that are exported. The fields are added both to the structure and its concrete signature.
- For `open`, the normal version (`open Simple.Path`) simply changes the typing environment, and does not introduce any runtime components into the current structure. However, the generalized version instead re-uses the same algorithm as `include`, with only a few minor changes. As with `include`, the fields of the opened structure are added to the current structure, and to its concrete signature. The twist is that we don't want to mark those fields as part of the inferred signature, so we mark the signature items as `Hidden`: https://github.com/ocaml/ocaml/blob/cbc5a0c807c77d8fb6d687e36f74af33e4c27bb5/typing/typemod.ml#L2470-L2488
In the end, we rely on the code for computing implicit constraints to compute the appropriate coercion. Note that we don't store anywhere which kind of algorithm we used; the only trace is in the `open_bound_items` field, which is the signature of the opened module in the general case but an empty signature for path-only `open`.
- Finally, in `Translmod`, we translate `Tstr_open` the same way as `Tstr_include`, except that to prevent generating worse code for the path-only case we use a test on `open_bound_idents` and the purity of the module expression to tell us when we can safely ignore the construction.

Maybe at some point I'll try to write some comments, but for now I'm happy with knowing that my code in #12782 is correct (and knowing why).
